src/QsCompiler/DataStructures/SyntaxExtensions.fs

```diff
@@ -372,52 +371,57 @@ let private TupleItems<'T when 'T :> ITuple> (arg: 'T): 'T list option =
     | :? SymbolTuple                as arg -> cast arg.TupleItems
     | _ -> InvalidOperationException("no extension provided for tuple matching of the given ITuple object") |> raise
 
-let (| Item | _ |) arg =         
+let (| Item | _ |) arg =
     match TupleItems arg with
     | Some [item] -> Some item
     | _ -> None
 
-let (| Tuple | _ |) arg =         
+let (| Tuple | _ |) arg =
     match TupleItems arg with
     | Some items when items.Length > 1 -> Some items
     | _ -> None
 
-let (| Missing | _ |) arg = 
+let (| Missing | _ |) arg =
     match TupleItems arg with
     | Some [] -> Some Missing
     | _ -> None
 
+// filter for type parameter resolution dictionaries
+
+[<Extension>]
+let FilterByOrigin (this : ImmutableDictionary<(QsQualifiedName * NonNullable<string>), ResolvedType>) origin =
```

Comment:
I think this function is a bit more generic:

```suggestion
let FilterByOrigin (this : KeyValuePair<_, _> seq) origin =
```

Although then it makes a little bit less sense as an extension method that returns a dictionary.

Reply:
The extension is supposed to operate specifically on type parameter resolutions dictionaries, filtering them out by their origins. The type of those dictionaries is `ImmutableDictionary<(QsQualifiedName * NonNullable<string>), ResolvedType>)`. Making it more permissive than it needs to be would have us supporting scenarios we don't intend to support right now. If in the future a situation comes up where we need to support different kinds of dictionaries with different type arguments, then the extension can be expanded at that point as needed without breaking code formed around it being specific to this type, but we would not be able to do the reverse without potentially breaking code.